frame/elections-phragmen/src/lib.rs

```diff
@@ -334,26 +407,58 @@ decl_module! {
 		///
 		/// A defunct voter is defined to be:
 		///   - a voter whose current submitted votes are all invalid. i.e. all of them are no
-		///     longer a candidate nor an active member.
+		///     longer a candidate nor an active member or a runner-up.
+		///
+		///
+		/// The origin must provide the number of current candidates and votes of the reported target 
+		/// for the purpose of accurate weight calculation.
 		///
 		/// # <weight>
-		/// #### State
-		/// Reads: O(NLogM) given M current candidates and N votes for `target`.
-		/// Writes: O(1)
+		/// No Base weight based on min square analysis.
+		/// Complexity of candidate_count: 1.755 us
+		/// Complexity of vote_count: 17.97us
+		/// State reads:
+		///  	- Voting(reporter)
+		///  	- Candidate.len()
+		///  	- Voting(Target)
+		///  	- Candidates, Members, RunnersUp (is_defunct_voter) [and Voting(Target) -- already counted]
+		/// State writes:
+		/// 	- Lock(reporter || target)
+		/// 	- AccountBalance(reporter) + AccountBalance(target)
+		/// 	- Voting(reporter || target)
+		/// Note: the db access is worse with respect to db, which is when the report is correct.
 		/// # </weight>
-		#[weight = 1_000_000_000]
-		fn report_defunct_voter(origin, target: <T::Lookup as StaticLookup>::Source) {
+		#[weight = FunctionOf(
+			|(defunct,): (&DefunctVoter<<T::Lookup as StaticLookup>::Source>,)|
+				defunct.candidate_count as Weight * (2 * WEIGHT_PER_MICROS) +
+				defunct.vote_count as Weight * (18 * WEIGHT_PER_MICROS) +
+				T::DbWeight::get().reads_writes(6, 4),
+			DispatchClass::Normal,
+			Pays::Yes,
+		)]
+		fn report_defunct_voter(
+			origin,
+			defunct: DefunctVoter<<T::Lookup as StaticLookup>::Source>,
+		) {
 			let reporter = ensure_signed(origin)?;
-			let target = T::Lookup::lookup(target)?;
+			let target = T::Lookup::lookup(defunct.who)?;
 
 			ensure!(reporter != target, Error::<T>::ReportSelf);
 			ensure!(Self::is_voter(&reporter), Error::<T>::MustBeVoter);
 
-			// Checking if someone is a candidate and a member here is O(LogN), making the whole
-			// function O(MLonN) with N candidates in total and M of them being voted by `target`.
-			// We could easily add another mapping to be able to check if someone is a candidate in
-			// `O(1)` but that would make the process of removing candidates at the end of each
-			// round slightly harder. Note that for now we have a bound of number of votes (`N`).
+			let DefunctVoter { candidate_count, vote_count, .. }  = defunct;
+
+			ensure!(
+				<Candidates<T>>::decode_len().unwrap_or(0) as u32 <= candidate_count,
+				Error::<T>::InvalidCandidateCount,
+			);
+			// Optimisation Note: We read this again in is_defunct_voter, still duplicate overlay
+			// read.
+			ensure!(
+				<Voting<T>>::get(&target).1.len() as u32 <= vote_count,
```

Comment:
checking this length involve decoding it, and decoding is considered `O(lenght)` but not sure of the real cost actually, in democracy I did swap the tuple element and implemented a manual decoding. https://github.com/paritytech/substrate/pull/5828#discussion_r419484688

If wanted we should release a new codec version with https://github.com/paritytech/parity-scale-codec/pull/209

Otherwise depending on the constaint here maybe it fine.

Reply:
Swapping will not work here, what will work is decoupling voting targets and balances. But it used to like this and we merged them because we always read these two together. 

For now I will leave it and have a note there to maybe improve later.

Reply:
actually I can re-org the functions in a way that I read the voters once and then pass them around. this should be a bit more expensive in-memory, but will prevent the need for multiple decoding. 

@shawntabrizi technically this might need a re-run :( but I think we will do this at the very end once more. So not a big deal for now.

Reply:
ok, also I see now that votes is of lenght 16 as maximum so fine.